src/colmap/estimators/bundle_adjustment.cc

```diff
@@ -848,6 +851,212 @@ void RigBundleAdjuster::ParameterizeCameraRigs(Reconstruction* reconstruction) {
+    const Options& prior_options,
+    const BundleAdjustmentConfig& config,
+    const std::unordered_map<image_t, PosePrior>& image_id_to_pose_prior)
+    : BundleAdjuster(options, config),
+      prior_options_(prior_options),
+      image_id_to_pose_prior_(image_id_to_pose_prior) {
+  setRansacMaxErrorFromPriorsCovariance();
+}
+
+bool PosePriorBundleAdjuster::Solve(Reconstruction* reconstruction) {
+  loss_function_ =
+      std::unique_ptr<ceres::LossFunction>(options_.CreateLossFunction());
+
+  if (prior_options_.use_robust_loss_on_prior_position) {
+    prior_loss_function_ = std::make_unique<ceres::CauchyLoss>(
+        prior_options_.prior_position_loss_scale);
+  }
+
+  // Initialize images' position w.r.t. priors with a rigid sim3D alignment
+  use_prior_position_ = Sim3DAlignment(reconstruction);
+
+  // Fix 7-DOFs of BA problem if not enough valid pose priors
+  if (!use_prior_position_) {
+    const std::vector<image_t>& reg_image_ids = reconstruction->RegImageIds();
+    config_.SetConstantCamPose(reg_image_ids[0]);
+    config_.SetConstantCamPositions(reg_image_ids[1], {0});
+  }
+
+  // Normalize the reconstruction to avoid any numerical instability BUT do not
+  // transform priors as they will be transformed when added to ceres::Problem
+  normalized_from_metric_ = reconstruction->Normalize(/*fixed_scale=*/true);
+
+  SetUpProblem(
+      reconstruction, loss_function_.get(), prior_loss_function_.get());
+
+  if (problem_->NumResiduals() == 0) {
+    return false;
+  }
+
+  ceres::Solver::Options solver_options =
+      SetUpSolverOptions(*problem_, options_.solver_options);
+
+  ceres::Solve(solver_options, problem_.get(), &summary_);
+
+  // Transform back the reconstruction to its coordinate system state
+  reconstruction->Transform(Inverse(normalized_from_metric_));
+
+  if (options_.print_summary || VLOG_IS_ON(1)) {
+    PrintSolverSummary(summary_, "Pose Prior Bundle adjustment report");
+  }
+
+  return true;
+}
+
+void PosePriorBundleAdjuster::SetUpProblem(
+    Reconstruction* reconstruction,
+    ceres::LossFunction* loss_function,
+    ceres::LossFunction* prior_loss_function) {
+  // Set up problem
+  // Warning: SetUpProblem must be called before AddPosePriorToProblem()
+  // Do not change order of instructions!
+  BundleAdjuster::SetUpProblem(reconstruction, loss_function_.get());
+
+  if (use_prior_position_) {
+    for (const auto& id_and_prior : image_id_to_pose_prior_) {
+      AddPosePriorToProblem(id_and_prior.first,
+                            id_and_prior.second,
+                            reconstruction,
+                            prior_loss_function_.get());
+    }
+  }
+}
+
+void PosePriorBundleAdjuster::AddPosePriorToProblem(
+    image_t image_id,
+    const PosePrior& prior,
+    Reconstruction* reconstruction,
+    ceres::LossFunction* prior_loss_function) {
+  if (!prior.IsValid() || !prior.IsCovarianceValid()) {
+    LOG(ERROR) << "Could not add prior for image #" << image_id << "\n";
```

Comment:
The newline is added by glog automatically.
```suggestion
    LOG(ERROR) << "Could not add prior for image #" << image_id;
```